### ISSUE TITLE
Stack doesn't rebuild enough when using Custom build-type

### ISSUE THREAD

**Issue by mboes** [commercialhaskell/stack#2040]
#### Steps to reproduce

Create a trivial stack project with the following Setup.hs (and setting `build-type: Custom`):

```Haskell
import Distribution.Simple

main = defaultMainWithHooks simpleUserHooks { postBuild = putStrLn "Hey!" }
```

Then `stack build` twice.

#### Expected

stack should run the preBuild/postBuild hooks each time `stack build` is called. In general these may be callouts to `make` or any other foreign build system, whose dependency graph is invisible to Cabal/Stack/GHC.

#### Actual

stack runs these hooks once only, unless Haskell source files change. Note that cabal-install does rerun these hooks every time. It's Stack's recompilation avoidance that seems to be the problem here.

**Comment by mboes** [comment]
One possible solution here could be to treat `build-type: Custom` specially, i.e. be less aggressive about recompilation avoidance in that particular case.

**Comment by mgsloan** [comment]
Yeah, special casing `built-type: Custom` for this makes sense to me.

**Comment by mboes** [comment]
Problem is, if package D has `build-type: Custom`, and the dependency chain in a multi-package project looks like A -> B -> C -> D, then the hooks will be rerun each time for D and presumably A, B, C become out-of-date and rebuild too? Likely no object files will be rebuilt, but the null build may take quite a bit more time.

What about another solution: take into account the `extra-source-files` field, and posit that `build-type: Custom` only ever needs to rerun if any of the sources files listed there (+ those files already taken into account) are newer. This is perhaps not correct in general, but should cover nearly every sane use case for using build hooks.

**Comment by mgsloan** [comment]
Yeah, though it seems to me like a strange behavior to toggle based on `build-type`.  Perhaps changes to  `extra-source-files` should always cause package dirtiness?  It's better to have false positives than false negatives for dirtiness checking.

**Comment by borsboom** [comment]
I think this is resolved by the merge of #2362. Please re-open if not.
